### ISSUE TITLE
Separated screen on/off rules support

### ISSUE THREAD

**Issue by baongoc124** [ukanth/afwall#492]
Could AFWall+ supports separated rules based on screen state (on/off)? That would be great since there're many apps I only want to have Internet connection when screen is on.

**Comment by Gitoffthelawn** [comment]
You know, that's a real interesting idea.  I have to agree: For most apps, I only want them to communicate when the screen is on.  I would take it even one step further: I only want them to communicate when the screen is on _and_ they are the foreground app.

**Comment by CHEF-KOCH** [comment]
I hope this is a joke, disable a firewall while screen is off, really? Do you do the same with your PC while you're idle/not on the keyboard? I only can pray you not do this. I see no argument in implementing this, because to the one simple thing, if you return you need to unlock your device and the screen goes back on ... so what's the deal then? 

I highly suggest that you ask the developer of the apps to implement it for several status like (only Wifi, only if unlocked and and and and) but I see no reasons for a firewall to monitor the status of screen sensor (which then coasts more battery/permissions btw). For such special cases Tasker also can kick-in because it already got that feature to monitor device status.

**Comment by baongoc124** [comment]
@CHEF-KOCH I think you misunderstood here. I want to have more strict rules when the screen is off, which means more apps cannot connect to Internet in that condition. Not the other way around.

**Comment by CHEF-KOCH** [comment]
Can you give me an real-world example with an app you want to restrict in this case, thanks.

**Comment by baongoc124** [comment]
Facebook for example. I don't care about notification when screen is off anyway. I know that it can be Greenified but it will kill the process and launch it again is slow.

**Comment by CHEF-KOCH** [comment]
Okay I understand you now, and it's in general possible but the problem is that apps like Facebook, WhatsApp, ... using gcm (Google Could Messaging)/Polling (as fallback) and you then have not only to block the app itself you also need to block gsm (google play-services) which provides the api/interface for gcm and sends the request. 

If you only want to block the notifications there are better solutions, if we talking about the traffic then it get's complicated because entirely blocking google play-services breaks a lot of stuff and not only for one app (in general).

**Comment by baongoc124** [comment]
You're right, @CHEF-KOCH . I actually want to block all traffics, not just notifications. Maybe we can disable push service of target app but that'll make the solution depends on other app. Looks like it stucks here.
Anyway, Facebook messenger also uses polling as fallback. I'm not sure about Facebook either.

**Comment by CHEF-KOCH** [comment]
I know that it's possible to _kill_ (or start) the activity via e.g. `adb shell am start -n <name-xyz>` so overall it would be possible to 'stop' certain stuff like gcm broadcast and others, e.g. 3C ToolBox can do this via GUI (Taskmanager - press on an app - touch from right to left to see activity/services/...) so it could be automatically done, the problem is that that would require an global database and a huge effort because you need to watch each deactivated activity and of course this could be abused to stop Googles Ads services or 3th party ads in general (which in this case is against Googles Play Store terms). Another problem is that some developers integrate everything in one activity, so killing this would FC the app immediately after app restart/open. 

Since I not saw any firewall app or Anti-Virus product which can do this (and is not against any law/term) I think we can mark this as 'wontfix' then - as long no one can offer me/us something to play with - or Google change (maybe in the future) something to allow such behavior.

**Comment by baongoc124** [comment]
Well, as I can see that we have no acceptable solution for this, it's 'wontfix' then. Thank you for your elaboration!
Btw, this feature I saw NetGuard has, and I read somewhere that they implement it by change set of rules when screen turn on/off. But it cannot block push notification as you said.

**Comment by CHEF-KOCH** [comment]
From what I know NetGuard works not per-app basis so this is pretty 'useless' because it blocks all the traffic, if we implementing this the first will whine that app xyz for background sync not works anymore and we better integrate an white-/backlist.

NetGuard (from what I know):
* Allows/Deny Wlan in general
* Allow/Deny Cellular network in general
* Allow WLan if screen is on 
* Allow Cellular Networks if screen is on 
* Block roaming 
* ...

For me that is no option because given reason, please also remember that the Android VPN APi suffers from more _problems_ as iptables (compared one-by-one). 


Sorry for disappoint you guys but everything is not really _optimal_. :disappointed:

**Comment by baongoc124** [comment]
:+1:  Perhaps this or many other limitations now (if has any) could be implemented as extra features that only can be activated if Xposed is available like Greenify does with some of their experimental features.
I tried Xposed once and its power is addictive. Cannot use any phone without it now. :-)

**Comment by CHEF-KOCH** [comment]
In latest 0.81 NetGuard release I found toggles for screen off/on on per-app basis, this sounds interesting then.I will look into the code and see if that is possible with iptables too but I guess you have to work with UID's separately, this is always bad because then it usually get race conditions (known iptables bug) especially if you have 100 or more apps and control each app separately. 

![images mobilism org](https://cloud.githubusercontent.com/assets/727123/12761253/79c7f3fc-c9ea-11e5-8b8b-adf4b880fa52.png).

So, it seems possible but I must first check how good this work because if it coasts more energy then it's a question if that's useful.

**Comment by Gitoffthelawn** [comment]
An option to block all (including GCM) internet traffic once the screen is off would definitely be useful.  Per app is a plus, but blocking all internet traffic would still be very useful.  It would save battery significantly.  Ideally, it would wait 10-30 seconds after the screen is off, to make sure you aren't going to turn it right back on.

This is a different compromise from automatically turning off WiFi / data when the screen is off.  When an app turns off WiFi / data, and then turns it back on, more battery is conserved, but credentials must be authenticated each time, which takes time (and a little power).  Blocking all internet traffic maintains the established connection, but just blocks the transmission/reception of data.  Thus, WiFi / data will be available immediately when the screen it turned back on.

**Comment by SkyWheel** [comment]
I've just logged in to create a feature request to block/enable network while screen off/on on per-app basis as it's realised in NetGuard.
With NetGuard this feature is extremely useful.
With AFWall+ I've created two separated profiles for different screen states. It's not very comfortable to switch them manually.

**Comment by someone-somenet-org** [comment]
Most likely not a wise idea to add. Resyncing when screen comes back on will most definitely eat more traffic than keeping an idle connection.
If you want to get rid of an app that possibly does problematic stuff, just uninstall it.

**Comment by Gitoffthelawn** [comment]
That is a big generalization without any evidence to support it.

Also, it's a gross oversimplification to "just uninstall it".

Every app can be considered "problematic", as an app with internet access can transmit or receive _any_ data.  Also, apps can work in conjunction with each other to transmit or receive data that a single app may not have access to.

**Comment by someone-somenet-org** [comment]
You need to be able to trust apps you have installed. If you cant, uninstall them. If you cannot trust any app... well throw away your device. This goes for all apps, not only apps with internet access. There are many more weaknesses on a rooted android, than internet access.

"Also, apps can work in conjunction with each other to transmit or receive data" cannot be solved by iptables. You will need an intent filter, which has nothing to do with iptables and should therefore be out of scope for this app.

This app is a **great iptables manager**, and imho should stay that way and not try to grow into a failing wannabe-all-security-solution.
Why not add an anti-virus part too? It has to do something with security too. Also some crypto-helper stuff, because well ... security has to do with crypto no? Also bundle tor for privacy!
... NOT

**Comment by SkyWheel** [comment]
@somenet 

It's not problem to trust an app or not to trust.
Actually some apps should not have access to internet while it's standby mode.
If some  app is trusted but I use it only while screen is on, this option could be very helpful.
You forgot very simple thing: not all mobile plans are unlimited.
As for me I need only three apps have access to internet on background: mail app and two messengers

**Comment by antomonte** [comment]
The app "NoRoot Data Firewall" allows easily to block depending on screen state (on/off) per app and much more comfortable than NetGuard.
It is highly comfortable to configure the app. I suggest to adapt this kind of feature. It's up to you.

**Comment by SkyWheel** [comment]
NoRoot Data Firewall is not FOSS. That's it.

**Comment by SkyWheel** [comment]
@ukanth , so, as far as I understood you won't implement the feature, please clarify.

**Comment by ukanth** [comment]
@SkyWheel , No, not yet. I'm working on few core level design changes. If that goes well, I might support this.

**Comment by GnaXi** [comment]
I would love to see this implemented!

**Comment by ildar** [comment]
Isn't it better you get an app catching on/off events (and other events you
like) and sending profile switch command (aka intent) to Af wall?

**Comment by ildar** [comment]
Easer (Auto operation performer) - https://f-droid.org/app/ryey.easer could
be a good candidate but needs your desired screen on/off event. Ask the
developer, he'll help

**Comment by Gitoffthelawn** [comment]
@ildar IIRC, switching profiles requires rewriting iptables, which is slow.

**Comment by SkyWheel** [comment]
@ildar, easer constantly crashes on my latest  nightly LineageOS build :cry:

**Comment by ildar** [comment]
it shouldn't. Please file it at https://github.com/renyuneyun/Easer/issues/

**Comment by ka223** [comment]
> @SkyWheel , No, not yet. I'm working on few core level design changes. If that goes well, I might support this.

@ukanth 
Any progress on this feature? 
Would be very useful.  Thank you for your work!

**Comment by toas-koas** [comment]
Is there any update?
